src/ring_theory/polynomial/homogeneous.lean

```diff
@@ -225,6 +237,158 @@ noncomputable theory
 open_locale classical
 open finset
 
+open_locale direct_sum
+
+/-- A version of `homogeneous_component` that maps into `homogeneous_submodule σ R n`. -/
+def homogeneous_component' [comm_semiring R] (n : ℕ) :
+  mv_polynomial σ R →ₗ[R] homogeneous_submodule σ R n :=
+let f := finsupp.restrict_dom R R {d : σ →₀ ℕ | ∑ i in d.support, d i = n} in
+(submodule.of_le $ (homogeneous_submodule_eq_finsupp_supported _ _ _).symm.le).comp f
+
+/-- Split a polynomial into a direct sum of homogeneous components. -/
+def to_homogeneous_components [comm_semiring R] :
+  mv_polynomial σ R →ₐ[R] (⨁ i, homogeneous_submodule σ R i) :=
+begin
+  refine add_monoid_algebra.lift _ _ _ _,
+  exact {
+    to_fun := λ d, direct_sum.of (λ i, homogeneous_submodule σ R i) (d.to_add.sum $ λ i x, x)
+      ⟨monomial d.to_add 1, is_homogeneous_monomial _ _ _ rfl⟩,
+    map_one' := rfl,
+    map_mul' := λ d₁ d₂, _, },
+  refine (dfinsupp.single_eq_of_sigma_eq (sigma.subtype_ext _ _)).trans
+    (direct_sum.of_mul_of _ _).symm; dsimp only [set_like.coe_ghas_mul, subtype.coe_mk, to_add_mul],
+  { exact finsupp.sum_add_index (λ _, rfl) (λ _ _ _, rfl) },
+  { rw [monomial_mul, mul_one], },
+end
+
+lemma to_homogeneous_components_monomial_one [comm_semiring R] (d : σ →₀ ℕ) :
+  to_homogeneous_components (monomial d (1 : R)) =
+    direct_sum.of (λ i, homogeneous_submodule σ R i) (d.sum $ λ i x, x)
+      ⟨monomial d 1, is_homogeneous_monomial _ _ _ rfl⟩ :=
+(add_monoid_algebra.lift_single _ _ _).trans (one_smul _ _)
+
+@[simp]
+lemma to_homogeneous_components_monomial [comm_semiring R] (d : σ →₀ ℕ) (r : R) :
+  to_homogeneous_components (monomial d r) =
+    direct_sum.of (λ i, homogeneous_submodule σ R i) (d.sum $ λ i x, x)
+      ⟨monomial d r, is_homogeneous_monomial _ _ _ rfl⟩ :=
+begin
+  have : monomial d (r • 1 : R) = r • monomial d 1 := (finsupp.smul_single _ _ _).symm,
+  rw [←mul_one r, ←smul_eq_mul],
+  simp_rw this,
+  rw [alg_hom.map_smul, to_homogeneous_components_monomial_one, ←direct_sum.of_smul],
+  refl,
+end
+
+@[simp]
+lemma to_homogeneous_components_X [comm_semiring R] (d : σ) :
+  to_homogeneous_components (mv_polynomial.X d : mv_polynomial σ R) =
```

Comment:
Is `mv_polynomial` not open? I think it's nice if we can just write (and read) `X`.